diff --git a/src/hev-socks5-session.c b/src/hev-socks5-session.c
index 6252f63..417e090 100644
--- a/src/hev-socks5-session.c
+++ b/src/hev-socks5-session.c
@@ -161,11 +161,15 @@ hev_socks5_session_udp_bind (HevSocks5Server *self, int sock,
         dst = &addr;
     }
 
-    res = connect (sock, (struct sockaddr *)dst, sizeof (struct sockaddr_in6));
-    if (res < 0)
-        return -1;
+    if (dst->sin6_port) {
+        res = connect (sock, (struct sockaddr *)dst, sizeof (struct sockaddr_in6));
+        if (res < 0)
+            return -1;
 
-    HEV_SOCKS5 (self)->udp_associated = !!dst->sin6_port;
+        HEV_SOCKS5 (self)->udp_associated = 1;
+    } else {
+        HEV_SOCKS5 (self)->udp_associated = 0;
+    }
 
     alen = sizeof (struct sockaddr_in6);
     res = getsockname (sock, (struct sockaddr *)src, &alen);
